### ISSUE TITLE
How do we attribute code contributions after continuous refactoring?

### ISSUE THREAD

**Issue by tingkai-mai** [nus-cs2103-AY2223S1/forum#413]
How do we attribute code contributions after continuous refactoring in RepoSense?

Following the usual PR workflows, our team performs a lot of refactoring, especially after a PR is merged.

For example:
* Person A works on a certain feature and creates a PR.
* Person B reviews and requests changes, and suggests changes as well to that PR
* Person A makes those changes (which can be argued to be attributed to Person B)
* Person B approves the PR and merges to master
* Person C refactors the code made by Person A and B

Another example:
A particular line can be made by three different people.  

As we observe from this example, the code evolves multiple times across different people. How to attribute the code to is made even murkier by constant refactoring, especially if entire files are deleted or renamed. 

In this case, how should we attribute the code contributions in different lines of code?

This may be related to #397 and #412.

**Comment by damithc** [comment]
@Ferusel at the end, you need to decide who should be accountable/responsible for, and take credit for, a class or a method.
It can be the person who edited it last, wrote it initially, who edited it most, who was in charge of that component/feature, etc. (the first one is the default RepoSense behavior but you can override it with `@@author` tags). The git blame feature can help you see who edited the code before ([example](https://github.com/se-edu/addressbook-level3/blame/master/build.gradle)). We expect the developers who edited the code to have the best sense of how much each person contributed to that code.

If you think it is not appropriate for any _one_ person to claim authorship of a particular piece of code, you can even not attribute that code to anyone, by using an empty `@@author` tag.  In that case, no one will get credit for that code, and no one will get penalized for code quality issues in that code.

**Comment by Echomo-Xinyu** [comment]
> the first one is the default RepoSense behavior but you can override it with `@@author` tags

Hi prof, could you give an example of how this tag can be used? For example, if ABC wants to claim the authorship of this method band, should ABC be our real name or GitHub name? how this JavaDoc should be written exactly?

```java
/**
 * Returns 1 all the time.
 * @@author: ABC
 */

```

**Comment by Echomo-Xinyu** [comment]
> > the first one is the default RepoSense behavior but you can override it with `@@author` tags

> Hi prof, could you give an example of how this tag can be used? For example, if ABC wants to claim the authorship of this
> method, should ABC be our real name or GitHub name? how this JavaDoc should be written exactly?

I have realised a link to the question I asked has been given in an earlier post. For those who want an answer like me, you can find the relevant information here: https://nus-cs2103-ay2223s1.github.io/website/admin/tools.html#tool-reposense-for-authorship-tracking

**Comment by damithc** [comment]
> I have realised a link to the question I asked has been given in an earlier post. For those who want an answer like me, you can find the relevant information here: https://nus-cs2103-ay2223s1.github.io/website/admin/tools.html#tool-reposense-for-authorship-tracking

@Echomo-Xinyu Yup. Don't mix up `@@author` tags used by RepoSense with `@author` tags used by JavaDoc.
